### ISSUE TITLE
Export packages of embedded JARs not packaged as OSGI-bundle using their Implementation-Version

### ISSUE THREAD

**Issue by xtracoder** [bndtools/bnd#1284]
In the use-case when I'm implementing some public API which uses in its interface classes from 3rd party JAR - i'm going to embed that JAR into my bundle.

I'm doing it via `instruction '-exportcontents=*' and including that JAR as resource. The problem is that JAR's packages are exported using version of the bundle, not version of the JAR.

Steps to reproduce are already described in another issue: 
https://github.com/bndtools/bnd/issues/1276#issuecomment-166602313 

While digging through Bnd code, I've found that versioning works correctly if the JAR being embedded is an OSGI bundle - in that case Export-Package versions defined in MANIFEST.mf are used. I think for non-OSGI JARs version from manifest's `Implementation-Version`  attribute should be used for all its packages (see here for details: https://docs.oracle.com/javase/7/docs/api/java/util/jar/Attributes.Name.html#IMPLEMENTATION_VERSION).

I've made an attempt to fix this. Please check content of that commit: https://github.com/xtracoder/bnd/commit/c7d4f62c8d0d42230464da56a6bd5bdb46139e08

**Comment by xtracoder** [comment]
Some bugs are fixed plus support for instruction '-jar-meta=<jar-name>.jar;version=1.2.3' are added in that commit https://github.com/xtracoder/bnd/commit/52492a05cd87a5d53f11d4314023f374b01774ce

**Comment by pkriens** [comment]
I do not think Implementation Version is well defined enough for this.

I will close this bug, I could live with a PR that provides a macro to get a manifest version from a Jar on the buildpath.

**Comment by xtracoder** [comment]
Re: "_I could live with a PR that provides a macro to get a manifest version from a Jar on the buildpath_"
Could you rephrase this so that regular people like me understand how to get this working?

**Comment by pkriens** [comment]
Well, you want to use a manifest header for a specific purpose that I do not think is well defined enough to make a default. So if there was a macro that would provide the contents of a manifest header of a jar:

            ${manifestheader;<file>;<header>}

Then you could instruct bnd to use that as version:

          Export-Package: com.example.*;version=${manifestheader;${jar};Implementation-Version}

**Comment by xtracoder** [comment]
Ok, now recommendation is clear, but it is actually of no help to me - I already have property which stores version of the target JAR and can use it as following (in Gradle):

    Export-Package: com.example.*;version=${versions.myJarVersion}

The problem with that approach appears in this live use-case:

1. I have a bundle, which introduces HTTP API to be used by others

2. Module is built on top of Apache HTTP libraries, in particular (notice - they have different version)

    httpclient-4.5.1.jar
    httpcore-4.4.4.jar
    httpmime-4.5.1.jar

3. All of them have common root package - org.apache.http, but sub-packages come from different JARs and via yours-recommended way I will have to manually introspect 3rd party libraries and list versions for roughly 20 packages. This is both time-consuming over-configuration and very good place for manual errors, especially when new versions of JARs in future will contain different set of packages.

At the moment the only configuration option for BndTools (compiled with my changes) I specify manually is - `-exportcontenst=*`, all other 'just works' .

PS: another thing I do not catch is why do you don't "think Implementation Version is well defined enough for this." Oracle's docs say that ["Implementation-Version manifest attribute is used for package versioning."](https://docs.oracle.com/javase/7/docs/api/java/util/jar/Attributes.Name.html#IMPLEMENTATION_VERSION). What is the ground to disregard that statement?

PPS: I admin that there can be developers that do not read docs and just do not put 'Implementation-Version' in published artifact. To workaround these (I bet very rare cases) - my version of BndTools add support for additional parameter - `-jarMeta=jarName;attr=value,jarName2;attr=value,`  - this way missing manifest attributes can be overriden from consumer side.

**Comment by pkriens** [comment]
The reason this is hard because what you want goes against the grain of OSGi and therefore bnd. We worked very hard to get a model where modules are separated by services defined in packages. This allows semantic versioning, baselining,  or what we feel is good software engineering. For example, if you activate baselining you would get an error when you forgot to update the version of the package when you update dependencies that cause a binary incompatible change. With package semantic versioning the bundle version != package version.

What you seem to want to do is take a shortcut because doing it right is too hard. That imho then seems to lose the whole value of using versions on packages. Implementation-Version has no syntax rules, it has not mandatory value, and has no requirement to be semantic. Or, it is just some increasing number with each release, which I think is easier to achieve with a simple timestam with Require-Bundle. And why is Specification-Version not better than Implementation-Version?

In this case a better solution might be to use a Contract. See [1] and [2]

Now this is largely my opinion and I do understand that bnd should also support enough rope for others.

A possible solution is therefore to create a macro that is only active during the package version augmentation that would identify the JAR. This macro could do something unique to a package. Because bnd knows where the package came from, we could lookup a manifest header from its source JAR:

          -exportcontents= *;version=${manifest;${@jar};Implementation-Version}

Would that work?

[1]: https://www.osgi.org/portable-java-contract-definitions/
[2]: http://blog.osgi.org/2013/08/osgi-contracts-wonkish.html

**Comment by xtracoder** [comment]
Ok, I've got your point regarding "_not well defined enough_" and yes, you are right about "_I want to do is take a shortcut because doing it right is too hard_". Actually, I'm far from being expert in OSGI and can't say what will is 'absolutely truly right' when I to need embed and export non-OSGI 3rd party library.

...

Regarding "Specification-Version vs. Implementation-Version" - the only thing we can typically do in current build systems - MAVEN/GRADLE/... - bind project dependency onto version of some JAR and that version is in 99.9% of cases is Implementation-Version. After that - resolution of package versions is done automatically (via Bnd for example), and source of the package version is in 99.9% of cases is `bundleVersion` manifest attribute, which in 99.9% of cases semantically is Implementation-Version of that JAR.  

Ok ... Specification-Version .. to my mind it can be used only in cases when JAR contains no implementation (i.e. code is distributed in pairs api.jar + api-impl.jar) - so when 'Implementation-Version' is missing in MANIFEST, then "Specification-Version" should be used.

...

      -exportcontents= *;version=${manifest;${@jar};Implementation-Version}

Would that work? - I can't tell for sure. 

- *When this macro will be executed?*

if it will be executed per each package, that might work. Originally I was thinking - it will be executed once withing BndTools.

- *What will happen if embedded JAR is OSGI bundle?*

What is the priority - version information in '-exportcontents' or package version in bundle MANIFEST.

- *Implementation-Version vs. Specification-Version*

From your question - it looks like there should be fallback mechanism from Implementation-Version to Specification-Version

I'm not sure all this can be sufficiently implemented in a macro.

**Comment by pkriens** [comment]
Yes it can :-) no in one, but in a number

But hearing you, I wonder if you should not just use a contract or require bundle since you will not get semantic versioning this way anyway?

**Comment by xtracoder** [comment]
I've looked through OSGI Contracts you've pointed out, but right now I do not have enough understanding to say if it is more suitable for us.

... and, regarding embedding 3rd party libs, ... taking into consideration all mentioned above, we here start thinking that the best approach will be repacking them into OSGI bundles one-by-one and distributing/installing in a side-by-side manner. This way 'unpatched' version of BndTools will work already plus there will be no conflicts of package exports (in case 2 bundles export same 3rd party library).

**Comment by pkriens** [comment]
I think this is the better approach ... These JARs are often not designed to work well with semantic versioning for packages and then trying to teach it from the outside is more pain than it is worth it. 

What does work well is to define a service interface designed with best OSGi practices using DS focused on YOUR usage. I.e. you do not try to make some super duper reusable component. Then you can often get away with a very small interface because all configuration (Config Admin) and life cycle stuff (DS) can be done INSIDE the bundle. 

As an example, I did this with mongodb. Instead of having their types all over my code I now have an extremely simple service based API.

This often leaves a surprisingly simple API. Since you maintain the API yourself, you can easily extend it over time in a semantic versioning way. That way I can often leave 99% of that kind of some open project's code inside my bundle. I often find that the system gets significantly simpler overall with this model.

However, that is not always a solution. It also has the drawback that for a lot of people it feels unintuitive; it flies in the face of reusability for many people. As always, design is about tradeoffs.

**Comment by xtracoder** [comment]
Thanks for more details, will keep it in mind.
